rust/cardano-blockchain-types/src/cip36/key_registration.rs

```diff
@@ -0,0 +1,368 @@
+//! <https://cips.cardano.org/cip/CIP-36>
+//! <https://github.com/cardano-foundation/CIPs/blob/master/CIP-0036/schema.cddl>
+
+use std::collections::HashSet;
+
+use anyhow::Context;
+use ed25519_dalek::VerifyingKey;
+use minicbor::{decode, Decode, Decoder};
+use pallas::ledger::addresses::{Address, ShelleyAddress};
+use strum::FromRepr;
+
+use super::voting_pk::VotingPubKey;
+use crate::utils::decode_helper::{decode_array_len, decode_bytes, decode_helper, decode_map_len};
+
+/// CIP-36 key registration - 61284
+///
+///
+/// ```cddl
+/// key_registration = {
+///     1 : [+delegation] / legacy_key_registration,
+///     2 : $stake_credential,
+///     3 : $payment_address,
+///     4 : $nonce,
+///     ? 5 : $voting_purpose .default 0
+//   }
+/// ```
+#[allow(clippy::module_name_repetitions)]
+#[derive(Clone, Default, Debug)]
+pub struct Cip36KeyRegistration {
+    /// Is this CIP36 or CIP15 format.
+    pub is_cip36: Option<bool>,
+    /// Voting public keys (called Delegations in the CIP-36 Spec).
+    /// Field 1 in the CIP-36 61284 Spec.
+    pub voting_pks: Vec<VotingPubKey>,
+    /// Stake public key to associate with the voting keys.
+    /// Field 2 in the CIP-36 61284 Spec.
+    pub stake_pk: VerifyingKey,
+    /// Payment Address to associate with the voting keys.
+    /// Field 3 in the CIP-36 61284 Spec.
+    pub payment_addr: Option<ShelleyAddress>,
+    /// Nonce (nonce that has been slot corrected).
+    /// Field 4 in the CIP-36 61284 Spec.
+    pub nonce: u64,
+    /// Registration Purpose (Always 0 for Catalyst).
+    /// Field 5 in the CIP-36 61284 Spec.
+    pub purpose: u64,
+    /// Raw nonce (nonce that has not had slot correction applied).
+    pub raw_nonce: u64,
+    /// Is payment address payable? (not a script)
+    pub is_payable: bool,
+}
+
+/// Enum of CIP36 registration (61284) with its associated unsigned integer key.
+#[derive(FromRepr, Debug, PartialEq)]
+#[repr(u16)]
+pub enum Cip36KeyRegistrationKeys {
+    /// Voting key.
+    VotingKey = 1,
+    /// Stake public key.
+    StakePk = 2,
+    /// Payment address.
+    PaymentAddr = 3,
+    /// Nonce.
+    Nonce = 4,
+    /// Purpose.
+    Purpose = 5,
+}
+
+impl Decode<'_, ()> for Cip36KeyRegistration {
+    fn decode(d: &mut Decoder, ctx: &mut ()) -> Result<Self, decode::Error> {
+        let map_len = decode_map_len(d, "CIP36 Key Registration")?;
+
+        let mut cip36_key_registration = Cip36KeyRegistration::default();
+
+        // Record of founded keys. Check for duplicate keys in the map
+        let mut found_keys: HashSet<u16> = HashSet::new();
+
+        for _ in 0..map_len {
+            let key: u16 = decode_helper(d, "key in CIP36 Key Registration", ctx)?;
+
+            if let Some(key) = Cip36KeyRegistrationKeys::from_repr(key) {
+                match key {
+                    Cip36KeyRegistrationKeys::VotingKey => {
+                        if !found_keys.insert(key as u16) {
+                            return Err(decode::Error::message(
```

Comment:
Do not fail fast, accumulate all possible errors in a validation_report like the original code intended to do.
This is to inform the user of all the reasons we failed to validate the metadata, and stopping on the first is unhelpful and frustrating UX.